docs/proposals/kubelet-systemd.md

```diff
@@ -0,0 +1,181 @@
+# Kubelet and systemd interaction
+
+**Author**: Derek Carr (@derekwaynecarr)
+
+**Status**: Proposed
+
+This document presents a specification for how the `kubelet` interfaces with `systemd`.
+
+## Motivation
+
+Many Linux distributions have either adopted, or plan to adopt `systemd` as their init system.
+
+This document describes a series of enhancements that should be made to the `kubelet` to better integrate with these distributions.
+
+## systemd units
+
+To help understand this proposal, we first provide a brief summary of `systemd` behavior.
+
+`systemd` manages a hierarchy of `slice`, `scope`, and `service` units.
+
+* `service` - application on the server that is launched by `systemd`; how it should start/stop; when it should be started; under what 
+circumstances it should be restarted; and any resource controls that should be applied to it.
+* `scope` - a process or group of processes which are not launched by `systemd` (i.e. fork), like a service, resource controls may be applied
+* `slice` - organizes a hierarchy in which `scope` and `service` units are placed.  a `slice` may contain `slice`, `scope`, or `service` units;
+processes are attached to `service` and `scope` units only, not to `slices`. The hierarchy is intended to be unified, meaning a process should
+only belong to a single leaf node.
+
+## cgroup hierarchy: orthogonal versus unified hierarchies and cgroupfs single writer
+
+Classical `cgroup` hierarchies were orthogonal per resource group controller, and a process may exist in different parts of the hierarchy.
+
+For example, a process `p1` could exist in each of the following at the same time:
+
+* `/sys/fs/cgroup/cpu/important/`
+* `/sys/fs/cgroup/memory/unimportant/`
+* `/sys/fs/cgroup/cpuacct/unimportant/`
+
+In addition, controllers for one resource group could depend on another in ways that were not always obvious.
+
+For example, the `cpu` controller depends on the `cpuacct` controller yet they were treated separately.
+
+Many found it confusing for a single process to belong to different nodes in the `cgroup` hierarchy across controllers.
+
+The Kernel direction for `cgroup` support is to move toward a unified `cgroup` hierarchy, where the per-controller hierarchies are eliminated in
+favor of hierarchies like the following:
+
+* `/sys/fs/cgroup/important/`
+* `/sys/fs/cgroup/unimportant/`
+
+In a unified hierarchy, a process should only belong to a single node in the `cgroup` tree.
+
+The Kernel direction for `cgroup` management is to promote a single-writer model rather than allowing multiple processes to independently
+write to parts of the file-system.
+
+## cgroup management with systemd and container runtimes
+
+By default, `systemd` defines three slices:
+
+* `system.slice` - contains system and application daemons
+* `user.slice` - user login related processes
+* `machine.slice` - virtual machines or containers that have init systems (not application space containers)
```

Comment:
@derekwaynecarr There are a few things going on here, and i don't really want to get derailed into a discussion about long term division of responsibilities between kubelet/systemd, etc.. just yet ;-)

But to try nail down the concepts for a minute: Kubernetes (really, the kubelet) is responsible for launching *pods*, and ideally it delegates everything involved therein to the container runtime. In the case of Docker, since a pod is an ad hoc assembly of different individual containers, the kubelet needs to manage the individual containers, and the pod can't be meaningfully said to exist outside of the context of the kubelet; but in the case of a container runtime like rkt, where the pod is launched as a first-class citizen, it _does_ have an external lifecycle and identity. In this circumstance, I think it's perfectly appropriate for pods and pod-like entities [1] to register themselves with systemd-machined and integrate with it wherever possible - including living under `machine.slice`. (Perhaps unsurprisingly, we do that by default for pods in rkt when systemd-machined is available.) This provides a unified, stable interface that anyone can use to query/manage various aspects of the pod (which is particularly handy with rkt since we don't expose any management daemon).

As to whether this is appropriate for **all** pods that Kubernetes launches: I would argue, yes, in the long term. But like I said, that derails into a much larger discussion about division of responsibilities between kubelet/systemd, etc. that I don't think we need to get into right here. I would, however, push back against the idea that `machine.slice` is for the domain of things like libvirt; to begin with, `system-nspawn` registers ALL containers it invokes with `systemd-machined` by default; and among other examples, dwalsh has long pushed for having Docker (and now runc) register with it too. AIUI in every case that registration creates a scope in `machine.slice`.

To get back to the original issue, though: I still feel strongly we should defer to systemd concepts and nomenclature when discussing these slices. I'm happy for you to add a caveat that "Kubelet does not directly launch pods in this slice" or similar if you want to be very explicit about the current behaviour.

[1]: I'm using "pod-like entities" to refer to the middle of the spectrum between "single process containers" and "virtual machines"; in [Lennart's parlance](http://www.freedesktop.org/wiki/Software/systemd/machined/), this is "OS containers", although I'm not hugely fond of that term